diff --git a/ffm.cpp b/ffm.cpp
index 8461152..536bd57 100644
--- a/ffm.cpp
+++ b/ffm.cpp
@@ -18,6 +18,7 @@ R: Precomputed scaling factor to make the 2-norm of each instance to be 1. len(R
 v: Value of each element in the problem
 */
 
+#include <cstdio>
 #pragma GCC diagnostic ignored "-Wunused-result" 
 #include <algorithm>
 #include <cmath>
@@ -33,6 +34,8 @@ v: Value of each element in the problem
 #include <vector>
 #include <cassert>
 #include <numeric>
+#include <mutex>
+#include <atomic>
 
 #if defined USESSE
 #include <pmmintrin.h>
@@ -529,7 +532,7 @@ void ffm_read_problem_to_disk(string txt_path, string bin_path) {
     }
 }
 
-ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param) {
+ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param, string log_path) {
 
     problem_on_disk tr(tr_path);
     problem_on_disk va(va_path);
@@ -558,10 +561,20 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
     cout << endl;
 
     Timer timer;
+    ofstream log{log_path};
+    mutex log_mtx;
+    if(log_path != "" && !log.good())
+        throw runtime_error("cannot open " + log_path);
+    if(log.good())
+        log << "epoch,global_step,epoch_accumulated_loss\n";
+        
+    int64_t step = 0, epoch = 0;
+    int64_t log_freq = max(1, tr.meta.l / 10);
 
     auto one_epoch = [&] (problem_on_disk &prob, bool do_update) {
 
         ffm_double loss = 0;
+        atomic<ffm_double> thread_local_losses[64]{};
 
         vector<ffm_int> outer_order(prob.meta.num_blocks);
         iota(outer_order.begin(), outer_order.end(), 0);
@@ -590,8 +603,26 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
                 ffm_double t = wTx(begin, end, r, model);
 
                 ffm_double expnyt = exp(-y*t);
+                auto instance_loss = log1p(expnyt);
 
-                loss += log1p(expnyt);
+                loss += instance_loss;
+
+                auto& tl_loss = thread_local_losses[omp_get_thread_num()];
+                auto prev = tl_loss.load(memory_order_relaxed);
+                tl_loss.store(prev + instance_loss, memory_order_relaxed);
+                if(do_update && ((step + ii) % log_freq == 0) && log.good())
+                {
+                    char buf[64];
+                    double synced_loss = 0;
+                    for(auto& tl_loss : thread_local_losses)
+                    {
+                        synced_loss += tl_loss.load(memory_order_relaxed);
+                    }
+
+                    snprintf(buf, sizeof(buf), "%ld,%ld,%f", epoch, step + ii, synced_loss);
+                    lock_guard<mutex> lck{log_mtx};
+                    log << buf << endl;
+                }
 
                 if(do_update) {
                    
@@ -600,7 +631,11 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
                     wTx(begin, end, r, model, kappa, param.eta, param.lambda, true);
                 }
             }
+            if(do_update)
+                step += l;
         }
+        if(do_update)
+            epoch += 1;
 
         return loss / prob.meta.l;
     };
diff --git a/ffm.h b/ffm.h
index fb7b8d7..a1e0bb3 100644
--- a/ffm.h
+++ b/ffm.h
@@ -42,7 +42,7 @@ void ffm_save_model(ffm_model &model, string path);
 
 ffm_model ffm_load_model(string path);
 
-ffm_model ffm_train_on_disk(string Tr_path, string Va_path, ffm_parameter param);
+ffm_model ffm_train_on_disk(string Tr_path, string Va_path, ffm_parameter param, string log_path);
 
 ffm_float ffm_predict(ffm_node *begin, ffm_node *end, ffm_model &model);
 
