diff --git a/ffm.cpp b/ffm.cpp
index 8461152..88a1723 100644
--- a/ffm.cpp
+++ b/ffm.cpp
@@ -18,6 +18,7 @@ R: Precomputed scaling factor to make the 2-norm of each instance to be 1. len(R
 v: Value of each element in the problem
 */
 
+#include <cstdio>
 #pragma GCC diagnostic ignored "-Wunused-result" 
 #include <algorithm>
 #include <cmath>
@@ -529,7 +530,7 @@ void ffm_read_problem_to_disk(string txt_path, string bin_path) {
     }
 }
 
-ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param) {
+ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param, string log_path) {
 
     problem_on_disk tr(tr_path);
     problem_on_disk va(va_path);
@@ -558,6 +559,8 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
     cout << endl;
 
     Timer timer;
+    ofstream log{log_path};
+    int64_t step = 0;
 
     auto one_epoch = [&] (problem_on_disk &prob, bool do_update) {
 
@@ -592,6 +595,12 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
                 ffm_double expnyt = exp(-y*t);
 
                 loss += log1p(expnyt);
+                if(log.good())
+                {
+                    char buf[64];
+                    snprintf(buf, sizeof(buf), "%ld,%f\n", step + ii, loss);
+                    log << buf;
+                }
 
                 if(do_update) {
                    
@@ -600,6 +609,7 @@ ffm_model ffm_train_on_disk(string tr_path, string va_path, ffm_parameter param)
                     wTx(begin, end, r, model, kappa, param.eta, param.lambda, true);
                 }
             }
+            step += l;
         }
 
         return loss / prob.meta.l;
diff --git a/ffm.h b/ffm.h
index fb7b8d7..a1e0bb3 100644
--- a/ffm.h
+++ b/ffm.h
@@ -42,7 +42,7 @@ void ffm_save_model(ffm_model &model, string path);
 
 ffm_model ffm_load_model(string path);
 
-ffm_model ffm_train_on_disk(string Tr_path, string Va_path, ffm_parameter param);
+ffm_model ffm_train_on_disk(string Tr_path, string Va_path, ffm_parameter param, string log_path);
 
 ffm_float ffm_predict(ffm_node *begin, ffm_node *end, ffm_model &model);
 
